upip.py: fail with ECONNREFUSED, if socket.getaddrinfo() returns IPV6 first.

awachtler opened the pull request:
While fetching packages from a local pypiserver (the official ones did work) I got this error 
message:

```Installing to: dwl/
Queue: ['micropython-iot-utils']
http://127.0.0.1:8080/micropython-iot-utils/json
Installing micropython-iot-utils 0.0.2 from http://localhost:8080/packages/micropython-iot-utils-0.0.2.tar.gz
http://localhost:8080/packages/micropython-iot-utils-0.0.2.tar.gz
Error installing 'micropython-iot-utils': [Errno 111] ECONNREFUSED, packages may be partially installed
```
It turned out, that usocket.getaddrinfo() returns in position 0 the IPV6 address and in position 1 the wanted IPV4
address.

```
http://localhost:8080/micropython-iot-utils/json
records getaddrinfo 2
ai:  (10, 1, 6, None, bytearray(b'\n\x00\x1f\x90\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00'))
ai:  (2, 1, 6, None, bytearray(b'\x02\x00\x1f\x90\x7f\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00'))
```

The patch simply filters the list for family-type usocket.AF_INET.

It is tested with the Unix-Port of Micropython, but since the changed code uses standard micropython language elements, no effects on the embedded platforms are expected.

I hope that this simple fix can make it to the master-branch. 

Best Regards, Axel

Commit:
added support for port number in upip.py

adding a port number other then 443 to a pypi url maybe needed if a local
server like devpi is used.

awachtler committed:
Merge branch 'micropython:master' into master

Commit:
fix ECONNREFUSED error, if getaddrinfo() delivers IPV6 first

dpgeorge commented on tools/upip.py:
```diff
@@ -134,6 +134,9 @@ def url_open(url):
         ai = usocket.getaddrinfo(host, port, 0, usocket.SOCK_STREAM)
+        ai = [_ai for _ai in ai if _ai[0] == usocket.AF_INET]
+        if len(ai) < 1:
+            fatal("No IPV4 address for host: '%s'" % host)
```
This new code should probably go outside the try/except, because the try is intended to protect against `getaddrinfo` errors.

Also, because this code runs on very constrained systems would be good to optimise it to not use a list comprehension (which creates a new function to run the comprehension, and creates a temporary list).  Could instead be a for loop something this:
```python
for a in ai:
    if a[0] == AF_INET:
        ai = a
        break
else:
    fatal(...)
```

awachtler replied:
Thanks @dpgeorge for the sugestions, I pushed a new commit to the branch and tested it before with the UNIX Port.

How do I handle the commit message warnings and code coverage issue from the automatic checks above?

Commit:
tools/upip.py: changed code for selecting AF_INET address

As suggest by @dpgeorge, the loop was moved out of try except block and
refactored to consume less memory ressources.

dpgeorge commented:
I just realised there's a much simpler way to filter IPv4, using the 3rd arg to `getaddrinfo`:
```python
usocket.getaddrinfo(host, port, usocket.AF_INET, usocket.SOCK_STREAM)
```

But then the real question is: why doesn't the IPv6 address work?  Applying this change would be going a bit backwards, because we do eventually want to support IPv6 everywhere, see eg #7563.

awachtler commented:
Yes I agree, forcing  IPV4 would be a step back.

I just verified with a fresh checkout of the original sources that the issue still exists, so we don't hunting ghosts (regarding #7563).
I also gave a try to:

```python
usocket.getaddrinfo(host, port, usocket.AF_INET, usocket.SOCK_STREAM)
```
instead of 

```python
usocket.getaddrinfo(host, port, 0, usocket.SOCK_STREAM)
```

This did work and I could install packages from my server.
I would propose this solution:

```python
# near line :19
debug = False
index_urls = ["https://micropython.org/pi", "https://pypi.org/pypi"]
+ preferred_address_family = usocket.AF_INET
 
# at line :136
-        ai = usocket.getaddrinfo(host, port, 0, usocket.SOCK_STREAM)
+        ai = usocket.getaddrinfo(host, port, preferred_address_family, usocket.SOCK_STREAM)
```

The default value for `preferred_address_family`could also be `0` and a user with the same issue can 
set `upip.preferred_address_family = usocket.AF_INET`, this would also be fine.

Commit:
tools/upip.py: added option to control getaddrinfo

if --afinet is given on command line or upip.preferred_address_family = usocket.AF_INET
is used in a script, then usocket.getaddrinfo() prefers IPV4 resolution. This might help
to prevent ECONNREFUSED error in case local DNS delivers an IPV6 in the first position
on the result list.

Commit:
tools/upip.py: removed empty line

awachtler commented:
In the last commits I did 
 * push the above described changes to the PR.
 * tested it in a script using "import upip" and via command-line
 * added an commandline-option --afinet which enables the IPV4 filtering.

A full command line looks now so, in order to prefer IPV4:

```bash
ports/unix/micropython -m tools.upip install -i http://gretel-2:8080 --debug --afinet -p /tmp/upip  micropython-iot-utils
```

and this is the script:

```python
import upip
upip.debug = True
upip.preferred_address_family = 2
upip.index_urls = ["http://gretel-2:8080"]
upip.install("micropython-iot-utils", "dwl")
```

dpgeorge commented:
Thanks, this is looking a lot better now.

But I still wonder:
1. why doesn't IPv6 work?
2. is it possible in your case to specify the IPv4 address explicitly at the `-i` argument to force IPv4 that way (instead of the server name)?

awachtler commented:
The IPV6 issue I have to check later this week, probably need some wireshark setup to track this down.

An IPV4 address in the URL did work. This was one of the first experiments I did make, when I was puzzled about ECONNREFUSED error. But I thought, that it is more comfortable to have upip.py doing  the name-resolution rather then 
type `nslookup gretel-2` and copy and paste the IPV4 address.

dpgeorge commented:
> The IPV6 issue I have to check later this week, probably need some wireshark setup to track this down.

It would be good to know the issue, because maybe that can be fixed and then IPv6 would work correctly.  That would likely be a better solution.

> An IPV4 address in the URL did work.

Ok, good to know that worked.

> But I thought, that it is more comfortable to have upip.py doing the name-resolution rather then
> type `nslookup gretel-2` and copy and paste the IPV4 address.

I agree it's more comfortable.  But we need to remember that `upip.py` runs on very constrained devices and we don't want to add code to it unless it's really necessary.

awachtler commented:
Hi, sorry for keep this PR pending, last days was a bit busy.
I made an experiment comparing Python 3.6 and micropython. It showed that my IPV6 config is wrong or the DNS telling bullshit,
the routing is broken. 
```
socket_tests_ipv6$ python test_ipv6_socket.py 
====================
(<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('2003:e7:c720:2200:f81a:eacf:a7af:545c', 8080, 0, 0))
ERROR: [Errno 113] No route to host
====================
(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('172.16.1.166', 8080))
OK:  HTTP/1.0 200 OK

socket_tests_ipv6$ ../micropython-aw/ports/unix/micropython test_ipv6_socket.py 
====================
(10, 1, 6, None, bytearray(b'\n\x00\x1f\x90\x00\x00\x00\x00 \x03\x00\xe7\xc7 "\x00\xf8\x1a\xea\xcf\xa7\xafT\\\x00\x00\x00\x00'))
ERROR: [Errno 113] EHOSTUNREACH
====================
(2, 1, 6, None, bytearray(b'\x02\x00\x1f\x90\xac\x10\x01\xa6\x00\x00\x00\x00\x00\x00\x00\x00'))
OK:  b'HTTP/1.0 200 OK\r\n'
``` 
The short answer in old-school-forums would be "Go fix your IPV6 configuration!".
But the user might be in a situation where [s]he either does not know how to do that, [s]he has no admin access
or simply has no time to dig into the particular issue. 

So would it be OK for upip.py to iterate over `ai` until we got an connection and leave the loop then?
 
Another  thing that puzzles me too is, I get now  EHOSTUNREACH instead of ECONNREFUSED from micropython.

Since I can't attach python files, I copy the code of my test script here

```python
try:
    import usocket as sck
    upy = True
except:
    import socket as sck
    upy = False

host = "gretel-2"
port = 8080
urlpath = ""

ai = sck.getaddrinfo(host, port, 0, sck.SOCK_STREAM)
for _ai in ai:
    print("="*20)
    print(_ai)
    try:
        _s = sck.socket(_ai[0], _ai[1], _ai[2])
        _s.connect(_ai[-1])
        msg = "GET /%s HTTP/1.0\r\nHost: %s:%s\r\n\r\n" % (urlpath, host, port)
        if upy:
            _s.write(msg)
            resp = _s.readline()
        else:
            _s.send(bytes(msg, "ascii"))
            resp = _s.recv(256).decode("ascii").split("\n")[0]
        print("OK: ", resp)
    except Exception as e:
        print("ERROR:", e)
        #raise(e)
```

Commit:
upip.py: loop over getaddrinfo()-list and try to connect

* The commit removes the previously added "preferred_address_family" and --afinet option.
* Instead of trying to connect always to first list entry, all returned address records
  are iterated and tried to connect to. With the first successfull connection, the loop
  is exited with break. If no connectiion can be established, a fatal() error terminates the
  script.

awachtler committed:
Merge branch 'micropython:master' into master

awachtler commented:
Hi, sorry for keep this PR pending, last days was a bit busy.
I made an experiment comparing Python 3.6 and micropython. It showed that my IPV6 config is wrong or the DNS telling bullshit,
the routing is broken. 
```
socket_tests_ipv6$ python test_ipv6_socket.py 
====================
(<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('2003:e7:c720:2200:f81a:eacf:a7af:545c', 8080, 0, 0))
ERROR: [Errno 113] No route to host
====================
(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('172.16.1.166', 8080))
OK:  HTTP/1.0 200 OK

socket_tests_ipv6$ ../micropython-aw/ports/unix/micropython test_ipv6_socket.py 
====================
(10, 1, 6, None, bytearray(b'\n\x00\x1f\x90\x00\x00\x00\x00 \x03\x00\xe7\xc7 "\x00\xf8\x1a\xea\xcf\xa7\xafT\\\x00\x00\x00\x00'))
ERROR: [Errno 113] EHOSTUNREACH
====================
(2, 1, 6, None, bytearray(b'\x02\x00\x1f\x90\xac\x10\x01\xa6\x00\x00\x00\x00\x00\x00\x00\x00'))
OK:  b'HTTP/1.0 200 OK\r\n'
``` 
The short answer in old-school-forums would be "Go fix your IPV6 configuration!".
But the user might be in a situation where [s]he either does not know how to do that, [s]he has no admin access
or simply has no time to dig into the particular issue. 

So would it be OK for upip.py to iterate over `ai` until we got an connection and leave the loop then?
 
Another  thing that puzzles me too is, I get now  EHOSTUNREACH instead of ECONNREFUSED from micropython.

Since I can't attach python files, I copy the code of my test script here

```python
try:
    import usocket as sck
    upy = True
except:
    import socket as sck
    upy = False

host = "gretel-2"
port = 8080
urlpath = ""

ai = sck.getaddrinfo(host, port, 0, sck.SOCK_STREAM)
for _ai in ai:
    print("="*20)
    print(_ai)
    try:
        _s = sck.socket(_ai[0], _ai[1], _ai[2])
        _s.connect(_ai[-1])
        msg = "GET /%s HTTP/1.0\r\nHost: %s:%s\r\n\r\n" % (urlpath, host, port)
        if upy:
            _s.write(msg)
            resp = _s.readline()
        else:
            _s.send(bytes(msg, "ascii"))
            resp = _s.recv(256).decode("ascii").split("\n")[0]
        print("OK: ", resp)
    except Exception as e:
        print("ERROR:", e)
        #raise(e)
```

awachtler commented:
Implemented and tested the above proposed iteration over getaddrinfo()-records in upip.py now. With the change we also get rid of the preferred_address_family / --afinet option.

dpgeorge commented:
If I understand correctly, a summary of the above discussion is: if IPv6 is not working correctly on a network then upip fails.  This is hard to debug and fix and it would be nice if upip would iterate through all possible IP addresses until one worked.

But I think it's much better to fix the network rather than make upip robust (because it's intended to run on small systems).

Also there is a simple workaround for the case of bad networks, or when you want to force IPv4/IPv6: use it's IP address instead of name.

So I'll close this PR.  But thanks anyway for your efforts!
